tendermint/src/amino_types/signature.rs

```diff
@@ -1,22 +0,0 @@
-use super::validate;
-use crate::{chain, consensus};
-use bytes::BufMut;
-use prost::EncodeError;
-use tendermint_proto::types::SignedMsgType;
-
-/// Amino messages which are signable within a Tendermint network
-pub trait SignableMsg {
-    /// Sign this message as bytes
-    fn sign_bytes<B: BufMut>(
-        &self,
-        chain_id: chain::Id,
-        sign_bytes: &mut B,
-    ) -> Result<bool, EncodeError>;
-
-    /// Set the Ed25519 signature on the underlying message
-    fn set_signature(&mut self, sig: &ed25519::Signature);
-    fn validate(&self) -> Result<(), validate::Error>;
-    fn consensus_state(&self) -> Option<consensus::State>;
-    fn height(&self) -> Option<i64>;
-    fn msg_type(&self) -> Option<SignedMsgType>;
-}
```

Comment:
Will there be an equivalent of this trait? Currently, some of the types have `to_signable_bytes`, so the trait implementations can just use that.
In tmkms, it's used for over different requests: https://github.com/iqlusioninc/tmkms/blob/develop/src/session.rs#L171

Or is it expected that the equivalent of this trait would be moved to tmkms?

Reply:
I'll be honest, I didn't see the need for a trait here. (Your link shows that this might be incorrect, but I didn't find that specific part of the KMS code before.)
Since the trait is only used in two types: Vote and Proposal, the trait implementation is a lot of extra code for very little benefit. (Unfortunately, traits are not interfaces.)

I'm going to go through that code to see what the expectations are. One option I would like to investigate is a message type enum that includes all types of messages. In that case the code can be replaced with a `match message {}` and call the functions in there. Since we're talking about two types of structs, this is not that overbearing. (If the number of structs can explode, then maybe the trait is a good call.)

If this doesn't pan out, we can always reintroduce the trait.

Reply:
Ah, I think I found it: https://github.com/iqlusioninc/tmkms/blob/3e3ae75fee35d1d81258ed61c286b2efb240be40/src/session.rs#L113

You can see that when a request is identified as a Vote or a Proposal, it calls the generic `.sign()` function to get the signable bytes. But if we already identified the structs, there's no need to generalize over them any more.

Instead of
```
Request::SignProposal(req) => self.sign(req)?,
```

it could be something like:
```
Request::SignProposal(req) => SignProposalRequest::try_from(req).unwrap().to_signable_bytes(),
```

where the try_from ensures that the incoming data is valid. (and errors out if not. be sure to handle the error instead of unwrapping)

This was both the trait and the generic `sign` function is removed and we got a cleaner understanding of what's happening when calling that code.

Does this way of thinking helps? Hoesnt question. I feel it's clearer to me but I understand it's a different concept than previous iterations.

Reply:
Extra note: I went further into the code and I see that the `req` is an amino_type (or as we call it in the new version, a Raw type). I have a feeling that it could be upgraded into the domaintype even and all checks of incoming data moved into the `read` function. The benefits of domain types are that there is a strict input check when creating the domain type and it can happen very early in the process. By the time the `.to_signable_bytes()` function has to be called, you might have a valid type already with no need to check for invalid data.

Reply:
I'll try to look more into tmkms, but yes, I think using the domain types may make the code simpler there

Reply:
I'll just say I'm not a fan of this trait either